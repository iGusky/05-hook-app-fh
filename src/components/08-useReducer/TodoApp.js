import React,{ useEffect, useReducer } from 'react';

import { TodoAdd } from './components/TodoAdd';
import { TodoList } from './components/TodoList';

import './styles.css';

import { todoReducer } from './todoReducer'

const initialState = [];

//Ayuda a establecer el estado incial del reducer
const init = () => {
    
    return JSON.parse( localStorage.getItem('todos') ) || []
    
    //* Uso antes del local Storage
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender react',
    //     done: false
    // }];
}



export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, initialState, init); 

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify( todos ));
    
    }, [todos])

    const handleDelete = ( todoId ) => {
        const action = {
            type: 'delete',
            payload: todoId,
        };
        dispatch( action );
    };

    const handleToggle = ( todoId ) => {
        const action = {
            type: 'toggle',
            payload: todoId,
        };
        dispatch( action );
    }
    
    const handleAddTodo = ( newTodo ) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div>
            <h1>Todo App ({ todos.length })</h1>
            <hr></hr>

            <div className = "row">
                <div className = "col-md-7">

                    <TodoList 
                        todos={ todos } 
                        handleDelete={ handleDelete } 
                        handleToggle={ handleToggle } 
                    />

                </div>
                <div className = " col-md-5">

                    <TodoAdd 
                        handleAddTodo={ handleAddTodo }
                    />

                </div>
            </div>
        </div>
    )
}
