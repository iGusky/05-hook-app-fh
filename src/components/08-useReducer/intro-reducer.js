
// * Este serÃ¡ el estado inicial
const tareaInitial = [{
    id: 1,
    todo: 'Comprar queso',
    done: false,
}];

const todoReducer = ( state = tareaInitial, action ) => {
    if(action?.type === 'agregar'){
        return [...state, action.payload];
    }

    return state;
};

let todos = todoReducer();

const nuevaTarea = {
    id: 2,
    todo: 'Comprar tortillas',
    done: false,
}
const action = {
    type: 'agregar',
    payload: nuevaTarea,
}

todos = todoReducer( todos, action); 

console.log(todos);