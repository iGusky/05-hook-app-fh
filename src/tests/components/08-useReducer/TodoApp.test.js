
import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';


describe('Pruebas en <TodoApp />', () => {

    const wrapper = shallow(<TodoApp />);
    Storage.prototype.setItem = jest.fn( () => { } );

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();    
    });

    test('debe agregar un TODO', () => {
        const wrapper = mount( <TodoApp />);

        act( () => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
        });

        expect( wrapper.find('h1').text().trim() ).toBe('Todo App (2)');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
    });
    
    test('debe de mandarse a llamar el HandleDelete', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);
        expect( wrapper.find('h1').text().trim() ).toBe('Todo App (0)');

    })
    
    
})
