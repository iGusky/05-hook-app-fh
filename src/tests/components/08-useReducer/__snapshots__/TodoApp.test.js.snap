// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-md-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className=" col-md-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
